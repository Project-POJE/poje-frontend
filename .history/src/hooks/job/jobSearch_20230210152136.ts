export function useJobSearchPrefetch(keyword: string) {
  //Todo 검색결과 프리패치 함수
}

export function useJobSearch(keyword: string) {
  //Todo 검색결과 가져오는 함수
}
