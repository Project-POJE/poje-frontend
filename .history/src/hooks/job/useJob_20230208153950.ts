export default function useJob() {
  return <div></div>;
}
