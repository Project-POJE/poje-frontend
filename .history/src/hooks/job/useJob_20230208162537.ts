import { useQuery } from '@tanstack/react-query';
interface UseJobType {
  job: string;
}
export default function useJob({ job }: UseJobType) {
  const {} = useQuery();
  return 123;
}
