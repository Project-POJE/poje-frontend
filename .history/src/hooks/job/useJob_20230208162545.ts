import { useQuery } from '@tanstack/react-query';
interface UseJobType {
  job: string;
}
export default function useJob({ job }: UseJobType) {
  //const { data } = useQuery();
  return 123;
}
