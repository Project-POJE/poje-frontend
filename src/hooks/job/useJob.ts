export default function useJob() {
  return 123;
}
