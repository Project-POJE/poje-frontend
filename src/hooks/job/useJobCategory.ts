import service from '@src/service';
import { useQuery } from '@tanstack/react-query';

export default function useJobCategory() {
  const {
    data: category = [],
    isLoading,
    isSuccess,
  } = useQuery(
    ['category'],
    () => {
      return service.job.getJobCategory();
    },
    {
      onError: (e: unknown) => {
        console.log('에러에요에어에어에어에어에어에어에어');
      },
    }
  );
  return { category, isLoading, isSuccess };
}
