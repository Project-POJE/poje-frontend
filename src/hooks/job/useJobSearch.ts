import { useCallback } from 'react';
export default function useJobSearch() {
  const prefetchJobSearch = useCallback((keyword: string) => {
    //Todo 검색결과 프리패치 함수
    console.log('프리패칭 수행중 ...');
  }, []);

  const getJobListByKeyword = useCallback((keyword: string) => {
    //Todo 검색결과 반환하기
  }, []);
  return { prefetchJobSearch, getJobListByKeyword };
}
