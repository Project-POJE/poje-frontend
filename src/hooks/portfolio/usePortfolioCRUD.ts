import useConfirmModal from '@src/hooks/common/useConfirmModal';
import { useNavigate } from 'react-router-dom';
import service from '@src/service';
import {
  CreatePortfolioTemplateRequest,
  DeletePortfolioItemRequest,
} from '@src/service/types/portfolio';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import useModal from '../common/useModal';

export default function usePortfolioCRUD() {
  const queryClient = useQueryClient();
  const { setModal } = useModal();
  const { setConfirmModal } = useConfirmModal();
  const navigate = useNavigate();
  const { mutate: createPortfolio } = useMutation(
    (data: CreatePortfolioTemplateRequest) =>
      service.portfolio.createPortfolioTemplate(data),
    {
      onSuccess: async (data) => {
        const {
          data: {
            result: { portfolioId },
          },
        } = data;
        queryClient.invalidateQueries(['portfolios']);
        queryClient.invalidateQueries(['portfoliosList']);
        queryClient.invalidateQueries(['myPortfolios']);
        setModal('포트폴리오 생성완료', () => {
          navigate(`/portfolio/${portfolioId}`);
        });
      },
      onError: (data: Error) => {
        console.log(data.message);
        setModal(data.message);
      },
    }
  );

  const { mutate: deletePortfolio } = useMutation(
    ({ portfolioId }: DeletePortfolioItemRequest) =>
      service.portfolio.deletePortfolio({ portfolioId }),
    {
      onSuccess: () => {
        queryClient.invalidateQueries(['portfolios']);
        queryClient.invalidateQueries(['portfoliosList']);
        queryClient.invalidateQueries(['myPortfolios']);
      },
    }
  );

  const onDeletePortfolio = (portfolioId: string) => {
    setConfirmModal('정말로 삭제 하시겠습니까?', () => {
      deletePortfolio({ portfolioId });
    });
  };

  return {
    createPortfolio,
    onDeletePortfolio,
  };
}
