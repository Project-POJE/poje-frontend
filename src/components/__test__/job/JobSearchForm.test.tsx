import JobSearchForm from '@src/components/home/JobPage/JobSearchForm';
import useJobCategory from '@src/hooks/job/useJobCategory';
import usePortfolioCRUD from '@src/hooks/portfolio/usePortfolioCRUD';
import { mockJobCategory } from '@src/test_util/mock/data/job';
import {
  addMockFunctionType,
  withAllContext,
  withMemoryRouter,
} from '@src/test_util/util';
import { render, screen } from '@testing-library/react';
import { Route } from 'react-router-dom';

jest.mock('@src/hooks/job/useJobCategory');
jest.mock('@src/hooks/portfolio/usePortfolioCRUD');

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => {
  const originModule =
    jest.requireActual<typeof import('react-router-dom')>('react-router-dom');
  return {
    ...originModule,
    useNavigate: () => mockUseNavigate,
  };
});

const mockUseJobCategory: jest.Mock = addMockFunctionType(useJobCategory);
const mockUsePortfolioCRUD: jest.Mock = addMockFunctionType(usePortfolioCRUD);

describe('JobSearchForm Component', () => {
  beforeEach(() => {
    mockUseJobCategory.mockImplementation(() => ({
      category: mockJobCategory,
    }));
  });

  afterEach(() => {
    mockUseJobCategory.mockReset();
    mockUsePortfolioCRUD.mockReset();
    mockUseNavigate.mockReset();
  });

  it('스냅샷 테스트 with param이 주어졌을 때', () => {
    const mockCreatePortfolio = jest.fn();
    const mockNavigate = jest.fn();

    mockUsePortfolioCRUD.mockImplementation(() => ({
      createPortfolio: mockCreatePortfolio,
    }));
    mockUseNavigate.mockImplementation(() => ({
      createPortfolio: mockNavigate,
    }));

    const { asFragment } = render(
      withAllContext(
        withMemoryRouter(
          <Route
            path='/job/:type/search/:keyword'
            element={<JobSearchForm />}
          />,
          {
            initialEntries: ['/job/개발자/search/개발자'],
          }
        )
      )
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('경로에 "전체"라는 타입이 없으면 포트폴리오 만들기 버튼 렌더링', () => {
    const mockCreatePortfolio = jest.fn();
    const mockNavigate = jest.fn();

    mockUsePortfolioCRUD.mockImplementation(() => ({
      createPortfolio: mockCreatePortfolio,
    }));
    mockUseNavigate.mockImplementation(() => ({
      createPortfolio: mockNavigate,
    }));

    render(
      withAllContext(
        withMemoryRouter(
          <Route
            path='/job/:type/search/:keyword'
            element={<JobSearchForm />}
          />,
          {
            initialEntries: ['/job/개발자/search/개발자'],
          }
        )
      )
    );

    expect(screen.getByText('포트폴리오 만들기')).toBeInTheDocument();
  });

  it('경로에 "전체"라는 타입이 있으면 포트폴리오 만들기 버튼 렌더링 안함 ❌', () => {
    const mockCreatePortfolio = jest.fn();
    const mockNavigate = jest.fn();

    mockUsePortfolioCRUD.mockImplementation(() => ({
      createPortfolio: mockCreatePortfolio,
    }));
    mockUseNavigate.mockImplementation(() => ({
      createPortfolio: mockNavigate,
    }));

    render(
      withAllContext(
        withMemoryRouter(
          <Route
            path='/job/:type/search/:keyword'
            element={<JobSearchForm />}
          />,
          {
            initialEntries: ['/job/전체/search/개발자'],
          }
        )
      )
    );

    expect(screen.queryByText('포트폴리오 만들기')).not.toBeInTheDocument();
	});
	
	it('검색어 입력 후 검색 함수 호출', () => {
    const mockCreatePortfolio = jest.fn();
    const mockNavigate = jest.fn();

    mockUsePortfolioCRUD.mockImplementation(() => ({
      createPortfolio: mockCreatePortfolio,
    }));
    mockUseNavigate.mockImplementation(() => ({
      createPortfolio: mockNavigate,
    }));

    render(
      withAllContext(
        withMemoryRouter(
          <Route
            path='/job/:type/search/:keyword'
            element={<JobSearchForm />}
          />,
          {
            initialEntries: ['/job/전체/search/개발자'],
          }
        )
      )
    );

    expect(screen.queryByText('포트폴리오 만들기')).not.toBeInTheDocument();
  });
});
