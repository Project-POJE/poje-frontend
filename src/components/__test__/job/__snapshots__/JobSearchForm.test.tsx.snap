// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobSearchForm Component 스냅샷 테스트 with param이 주어졌을 때 1`] = `
<DocumentFragment>
  <form
    class="sc-bcXHqe jBFDOT"
  >
    <div
      class="sc-gswNZR bGIUVR"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          search-icon
        </title>
        <path
          d="M19.023 16.977a35.13 35.13 0 0 1-1.367-1.384c-.372-.378-.596-.653-.596-.653l-2.8-1.337A6.962 6.962 0 0 0 16 9c0-3.859-3.14-7-7-7S2 5.141 2 9s3.14 7 7 7c1.763 0 3.37-.66 4.603-1.739l1.337 2.8s.275.224.653.596c.387.363.896.854 1.384 1.367l1.358 1.392.604.646 2.121-2.121-.646-.604c-.379-.372-.885-.866-1.391-1.36zM9 14c-2.757 0-5-2.243-5-5s2.243-5 5-5 5 2.243 5 5-2.243 5-5 5z"
        />
      </svg>
      <input
        class="sc-dkrFOg iJAEwI"
        type="text"
        value="개발자"
      />
      <select
        class="sc-jSUZER dVFesh"
        name="jobCategory"
      >
        <option
          value="개발자"
        >
          개발자
        </option>
        <option
          value="디자이너"
        >
          디자이너
        </option>
        <option
          value="유승윤"
        >
          유승윤
        </option>
      </select>
      <button
        class="sc-hLBbgP bmDMKx"
        title="search-btn"
      />
    </div>
    <button
      class="sc-eDvSVe OjWVY"
      type="button"
    >
      포트폴리오 만들기
    </button>
  </form>
</DocumentFragment>
`;
